### PR TITLE
#4148: Add direct and interleaved ring gather tests between devices in a system with multiple WHs

### PR DESCRIPTION
**Pull request by tt-aho**

### CODE REVIEW

File: tt_metal/hw/inc/ethernet/dataflow_api.h [hunk 1]
```diff
@@ -65,6 +71,155 @@ void check_and_context_switch() {
     // proceed
 }
 
+/**
```
Comment (aliuTT) [inline tt_metal/hw/inc/ethernet/dataflow_api.h:74]: Do we need to copy every noc api to a `eth_noc` version? I would prefer if we only made the copy where we needed to add `risc_context_switch()` for functionality. That way if things improve in the noc api, we benefit here as well.

File: tests/tt_metal/tt_metal/unit_tests/multichip/ring_gather_kernels.cpp [hunk 2]
```diff
@@ -0,0 +1,483 @@
...
+        }
+    }
+    return {};
+}
+
+std::vector<Device*> get_device_ring(std::vector<tt::tt_metal::Device*> devices) {
+    std::vector<std::vector<int>> adj(devices.size(), std::vector<int>(devices.size(), 0));
+    for (uint32_t i = 0; i < devices.size(); ++i) {
+        const auto& device = devices[i];
+        for (const auto& connected_device_id : device->get_ethernet_connected_chip_ids()) {
```
Comment (aliuTT) [inline tests/tt_metal/tt_metal/unit_tests/multichip/ring_gather_kernels.cpp:83]: FYI, I made some changes to `get_ethernet_connected_chip_ids` to return a map of connect chips with their connected eth cores too. That PR will go in later than this one though

File: tt_metal/hw/inc/ethernet/dataflow_api.h (hunk 1 above)
Comment (tt-aho) [reply tt_metal/hw/inc/ethernet/dataflow_api.h:74]: Makes sense, will update

Comment (tt-aho) [reply tt_metal/hw/inc/ethernet/dataflow_api.h:68]: Updated as per discussion to include the regular dataflow api, so we do not need to duplicate the functions in this file, only the ones which need context switch are added.

**Pull request merged**